### PR TITLE
Set up testing and module path aliases

### PR DESCRIPTION
**Pull request by joannerd**

### CODE REVIEW

File: pages/api/hakkasan.ts [hunk 1]
```diff
@@ -18,11 +18,12 @@ const handler = async (
     case 'GET': {
       try {
         if (process.env.NODE_ENV !== 'production') {
-          await delay(2000);
-          res.status(200).send({
-            artists,
-            events,
-            venues,
+          await delay(1000, () => {
```
Comment (andweeb) [inline pages/api/hakkasan.ts:21]: Is this callback necessary? If we're `await`ing on some delay, wouldn't it be cleaner simply to make the `res.send` call after the `await` instead of passing it wrapped in an anonymous function?

File: test/hooks/useFetch.test.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    };
+
+    global.fetch = generateMockFetch({
+      json: () => Promise.resolve(expected),
+    });
+
+    const { result } = renderHook(() => useFetch<HakkasanResponse>(endpoint));
+
+    await act(async () => {
+      await result.current.fetchData();
```
Comment (andweeb) [inline test/hooks/useFetch.test.ts:25]: You can pass in `result.current.fetchData` directly as an argument to `act()` here and in [line 41](https://github.com/joannerd/hakkasan/pull/3/files#diff-9544fb45add65fa9f45714818f0dc8496eae7372c61a297779fe2aebb1265c1dR41). It might be good to keep in mind that creating inline `async` arrow functions when you don't need to leads to unnecessary overhead for the TS compiler.

For example, the compiled output of this inline async arrow function looks like this:
```js
    return __awaiter(this, void 0, void 0, function* () {
        yield act(() => __awaiter(this, void 0, void 0, function* () {
            yield fetchData();
        }));
    });
```

... and passing the function directly looks like this:
```js
    return __awaiter(this, void 0, void 0, function* () {
        yield act(fetchData);
    });
```

You can see the compiler intersperses the callback through an `awaiter` wrapper when it doesn't actually need to. In other words best practice would be to avoid using `async`/`await` when there's nothing to actually `await` for.

You can see the full TS Playground example [here](https://www.typescriptlang.org/play?target=2#code/IYZwngdgxgBAZgV2gFwJYHsLwKbKgCwBFhlgAKASgC4YAFAJ3QFtURsAeAN3VQBMA+GAG8AUDHEwI2AO51GLNmXrYQ6ADadsMALyC2yACqom2dAmRlKOwctUbsZJL2xxUU3hQA0MAIwAGAIoKEQBfERFEFAwsYCgLKAAjGitdOWZWDm4+fgphMQllZAR6LETKAG5Q8NBIWEi46JhkFWQfK1EJGGBpYFRkLriyGugYFMEOzq6evpw8IhJyCkrOkKWqkWG6pAbMJpaAJnb88W7e-tiLOFwCYlI1kKA).

File: test/hooks/useFetch.test.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    expect(result.current.data).toEqual(expected);
+  });
+
+  test('should set errors', async () => {
+    const endpoint = '/api/hakkasan?ref=events';
+    const expectedError = 'response.json is not a function';
+
+    global.fetch = generateMockFetch({
+      json: undefined,
+    });
```
Comment (andweeb) [inline test/hooks/useFetch.test.ts:37]: It would be better to mock the actual `json` method to reject with our own test error so that we don't rely on some message text from an error we don't control. For instance, if we were to update the Node version and they change the error message for when some non-function tries to be invoked, this test would fail when it shouldn't.

```suggestion
    const expectedError = 'response.json is not a function';

    global.fetch = generateMockFetch({
      json: () => Promise.reject(new Error(expectedError)),
    });
```

File: test/lib/hakkasan.test.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+describe('parseHakkasanData(text)', () => {
+  test('should parse the argument of retrieveJSONP from Hakkasan JSON data', () => {
+    const data = { message: 'Test data' };
+    const dataText = JSON.stringify(data);
+    const text = `retrieveJSONP(${dataText})`;
+    const result = parseHakkasanData(text);
+    expect(result).toEqual(data);
+  });
+
+  test('should catch and rethrow errors', () => {
```
Comment (andweeb) [inline test/lib/hakkasan.test.ts:18]: A couple things here!
1. The test assertion seems like it's only testing the error case in which no argument is passed into the function, which wouldn't actually be likely due to the fact that we're using TypeScript.
2. I'd say we can improve this test to be more specific in which _types_ of specific errors can be caught within the function with varying different types of text.
3. I'd recommend against creating tests as simple/generic as this as it doesn't exactly test much and give any hints as to what errors would (or should) be expected depending on the input.

File: test/lib/hakkasan.test.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    expect(parseHakkasanData).toThrowError();
+  });
+});
+
+describe('fetchHakkasanByRef(ref)', () => {
+  test('should not fetch Hakkasan data with an invalid ref', async () => {
```
Comment (andweeb) [inline test/lib/hakkasan.test.ts:24]: Hmm I'm thinking we might be able to remove this test case entirely by simply changing the typing of the `ref` argument of the function to be more accurately a string union of either `artists`, `events`, or `venues`.

With that we'd also be able to remove the `VALID_HAKKSAN_REFS` constant.

**Review by andweeb — CHANGES_REQUESTED** [review]

File: pages/api/hakkasan.ts (hunk 1 above)
Comment (joannerd) [reply pages/api/hakkasan.ts:21]: Hmm, this refactor was made so that it was easier to test.

File: test/hooks/useFetch.test.ts (hunk 2 above)
Comment (joannerd) [reply test/hooks/useFetch.test.ts:25]: Ooooh thank you!!

File: test/hooks/useFetch.test.ts (hunk 3 above)
Comment (joannerd) [reply test/hooks/useFetch.test.ts:37]: Good point! Thank you! :D

File: test/lib/hakkasan.test.ts (hunk 5 above)
Comment (joannerd) [reply test/lib/hakkasan.test.ts:24]: Ooh yes, that sounds like a cleaner approach!

File: test/lib/hakkasan.test.ts (hunk 4 above)
Comment (joannerd) [reply test/lib/hakkasan.test.ts]: Good point! Updated the test with a more realistic input (string input without a `retrieveJSONP()` invocation).

File: pages/api/hakkasan.ts (hunk 1 above)
Comment (andweeb) [reply pages/api/hakkasan.ts:21]: I'd prefer a cleaner API over easier unit tests, especially here since requiring a callback introduces further nesting and overhead when reading the code.

If we were to redefine the delay function back to something like this:
```ts
export const delay = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(() => resolve(), ms));
```

We can test delayed code execution using a Promise chain to call a mock function after a delay, advancing the time, and resolving all jobs in the [internal queue `PromiseJobs`](https://javascript.info/microtask-queue) and _then_ testing whether a function has been called:
```ts
  test('calls code after the given ms', async () => {
    const callAfterTimeout = jest.fn();
    actualDelay(2000).then(callAfterTimeout);        // create a promise chain to resolve after 2000ms

    jest.advanceTimersByTime(2000);                  // advance time by 2000ms
    await Promise.resolve();                         // run all pending callbacks in PromiseJobs
    expect(callAfterTimeout).toHaveBeenCalled();     // code should have been called
  });
```

We can also tweak this test to make it a negative test:
```ts
  test('does not call code before the given ms', async () => {
    const callAfterTimeout = jest.fn();
    actualDelay(2000).then(callAfterTimeout);        // create a promise chain to resolve after 2000ms

    jest.advanceTimersByTime(1000);                  // advance time by only 1000ms
    await Promise.resolve();                         // run all pending callbacks in PromiseJobs
    expect(callAfterTimeout).not.toHaveBeenCalled(); // code should not have been called
  });
```

**Review by andweeb — CHANGES_REQUESTED** [review]

**Review by andweeb — APPROVED** [review]
LGTM 💯
